src/main/java/org/jitsi/videobridge/ConferenceSpeechActivity.java

```diff
@@ -40,261 +41,135 @@
+                        .filter(e -> id.equals(e.getID()))
+                        .findFirst().orElse(null);
+            // Move this endpoint to the top of our sorted list
+            if (!endpointsBySpeechActivity.remove(endpoint))
             {
-                logger.warn("Unable to find endpoint corresponding to "
-                    + "active speaker SSRC " + ssrc);
+                logger.warn("Got active speaker notification for an unknown endpoint: " + id + ", ignoring");
                 return;
             }
+            endpointsBySpeechActivity.add(0, endpoint);
 
-            synchronized (syncRoot)
-            {
-                // Move this endpoint to the top of our sorted list
-                if (!endpoints.remove(endpoint))
-                {
-                    logger.warn("Got active speaker notification for an unknown"
-                            + " endpoint (ssrc: " + ssrc + ", epId "
-                            + endpoint.getID() + ")! Ignoring");
-                    return;
-                }
-                endpoints.add(0, endpoint);
+            endpointListChanged = updateLastNEndpoints();
+        }
 
-                TaskPools.IO_POOL.submit(() ->
-                {
-                    if (conference != null)
-                    {
-                        conference.dominantSpeakerChanged();
-                    }
-                });
+        TaskPools.IO_POOL.submit(() -> {
+            listener.dominantSpeakerChanged();
+            if (endpointListChanged)
+            {
+                listener.lastNEndpointsChanged();
             }
-        }
+        });
     }
 
     /**
-     * Retrieves a JSON representation of the dominant speaker
-     * for the purposes of the REST API of Videobridge.
-     *
-     * @return a <tt>JSONObject</tt> which represents
-     * <tt>dominantSpeakerIdentification</tt> for the purposes of the REST API
-     * of Videobridge
+     * Re-calculates the list of endpoints in LastN order ({@link #endpointsInLastNOrder}) based on the speech activity
+     * and video availability.
      */
-    public JSONObject doGetDominantSpeakerIdentificationJSON()
+    boolean updateLastNEndpoints()
     {
-        DominantSpeakerIdentification dominantSpeakerIdentification
-            = this.dominantSpeakerIdentification;
-        JSONObject jsonObject;
-
-        if (dominantSpeakerIdentification == null)
-        {
-            // We do not know how to represent ActiveSpeakerDetector at the time
-            // of this writing, we know how to represent
-            // DominantSpeakerIdentification only.
-            jsonObject = null;
-        }
-        else
+        synchronized (syncRoot)
         {
-            Conference conference = this.conference;
+            Map<Boolean, List<AbstractEndpoint>> bySendingVideo
+                    = endpointsBySpeechActivity.stream()
+                    .filter(Objects::nonNull)
```

Comment:
How would we get null objects on this list?

Reply:
We can't as far as I can tell. It would be an illegal state. I removed the filter.